lib/shopify-cli/monorail.rb

```diff
@@ -0,0 +1,109 @@
+require 'json'
+require 'securerandom'
+require 'time'
+require 'timeout'
+require 'net/http'
+
+module ShopifyCli
+  module Monorail
+    autoload :Log, 'shopify-cli/monorail/log'
+
+    ENDPOINT_URI = URI.parse('https://monorail-edge.shopifycloud.com/v1/produce')
+
+    MonorailError = Class.new(StandardError)
+
+    class NilWriter
+      def write(*args); end
+    end
+
+    class << self
+      attr_writer :monorail, :events
+
+      def log
+        return @monorail if @monorail
+        prompt_for_consent
+        writable = if enabled? && consented?
+          events
+        else
+          NilWriter.new
+        end
+        @monorail = ShopifyCli::Monorail::Log.new(writable: writable)
+      end
+
+      def send_events
+        return unless enabled?
+        return unless consented?
+        new_events = events.tail(200).select do |line|
+          event = JSON.parse(line, symbolize_names: true)
+          Time.parse(event[:payload][:timestamp]) > mtime
+        end
+
+        new_events.reverse.each do |line|
+          event = JSON.parse(line, symbolize_names: true)
+          produce(event)
+          File.write(ShopifyCli::EVENTS_MTIME, event[:payload][:timestamp])
+        end
+      end
+
+      def produce(event, num_retries: 3)
+        headers = {
+          'Content-Type': 'application/json; charset=utf-8',
+          'X-Monorail-Edge-Event-Created-At-Ms': Time.parse(event[:payload][:timestamp]).to_i.to_s,
+          'X-Monorail-Edge-Event-Sent-At-Ms': Time.now.utc.to_i.to_s,
+        }
+        do_post(headers, JSON.dump(event), num_retries)
+      rescue Exception => e # rubocop:disable Lint/RescueException
+        ShopifyCli::Logger.error(name) { "Unexpected error when posting #{event.inspect}: #{e.inspect}" }
+      end
+
+      private
+
+      def prompt_for_consent
+        return unless enabled?
+        return if ShopifyCli::Config.get_section('analytics').key?('enabled')
+        msg = <<~MSG
+          Would you like to enable anonymous usage reporting?
+          We collect data about which commands you run and report errors to help us improve this tool.
```

Comment:
Hey @gfscott any suggestions about this prompt?

Reply:
I like it. How about just a tad more detail:

> Would you like to enable anonymous usage reporting?
> If you select “Yes”, we’ll collect data about which commands you use and which errors you encounter. We don’t collect personally identifiable information. Sharing this anonymous data helps Shopify improve this tool.

I like that we're being clear about what exactly will be collected; are "commands" and "errors" the only two things we log?

Reply:
Awesome, thank you. Yes commands and errors are it so far.

Reply:
Is there any risk of error output containing potentially identifiable information? It's kind of hard to know what might get thrown, correct?

Reply:
Yeah, I'm not sure we can guarantee that right now, but we could make this smarter in the future, so I may leave that clause out.

Reply:
Yeah, in that case I'd just lose the sentence about PII.